core-rust/node-common/src/scheduler.rs

```diff
@@ -62,190 +62,216 @@
+    name: String,
+    running_task_tracker: Option<&'a RunningTaskTracker>,
+    metrics: Option<SchedulerMetrics>,
 }
 
-/// Forward/delegate implementation of [`CancellableToken`] for Tokio's [`CancellationToken`]
-impl CancellableToken for CancellationToken {
-    fn cancel(&self) {
-        self.cancel();
+impl<'a> Scheduler<'a> {
+    /// Creates a new scheduler instance using the given tokio `runtime`.
+    /// The given `base_name` simply becomes the first segment of the constructed tasks' names' (see
+    /// [`Self::named()`]).
+    pub fn new(runtime: &'a Runtime, base_name: impl Display) -> Self {
+        Self {
+            runtime: Some(runtime),
+            name: base_name.to_string(),
+            running_task_tracker: None,
+            metrics: None,
+        }
     }
-}
 
-pub struct NoopCancellationFuture;
-
-impl Future for NoopCancellationFuture {
-    type Output = ();
-
-    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<()> {
-        Poll::Ready(())
+    /// Creates a mock instance for testing purposes.
+    /// The "scheduled" tasks will not actually be started.
+    pub fn for_testing() -> Self {
+        Self {
+            runtime: None,
+            name: "testing".to_string(),
+            running_task_tracker: None,
+            metrics: None,
+        }
     }
-}
-
-/// Noop implementation of the [`CancellableToken`]
-pub struct NoopCancellableToken;
 
-impl CancellableToken for NoopCancellableToken {
-    fn cancel(&self) {}
-}
-
-/// A utility wrapper around a [`CancellableToken`] which will
-/// cancel the task when dropped.
-pub struct CancelOnDrop<CT: CancellableToken> {
-    pub cancellation_token: CT,
-}
+    /// Appends another segment to the dot-separated names of tasks *created from this point on*.
+    /// The names are used only for error-surfacing and metrics purposes.
+    pub fn named(&self, segment: impl Display) -> Self {
+        let mut derived = self.clone();
+        derived.name = format!("{}.{}", derived.name, segment);
+        derived
+    }
 
-impl<CT: CancellableToken> CancelOnDrop<CT> {
-    pub fn new(cancellation_token: CT) -> Self {
-        Self { cancellation_token }
+    /// Configures the auto-tracking of the tasks *created from this point on*.
+    /// Every [`RunningTask`] returned by the task-starting method will also be reported to the
+    /// given tracker (which may be easier for the application's boot-up logic to keep alive).
+    pub fn track_running_tasks(&self, running_task_tracker: &'a RunningTaskTracker) -> Self {
+        let mut derived = self.clone();
+        derived.running_task_tracker = Some(running_task_tracker);
+        derived
     }
-}
 
-impl<CT: CancellableToken> Drop for CancelOnDrop<CT> {
-    fn drop(&mut self) {
-        self.cancellation_token.cancel();
+    /// Configures the metrics collection for tasks *created from this point on*.
+    pub fn measured(&self, metrics_registry: &Registry) -> Self {
```

Comment:
_(note from author)_

one extra capability I imagine would be `.resume_after_panic()`

(I tested, and currently, a panic inside a task run is logged [by the default panic handler] and the task dies; maybe we would prefer that it resumes execution on the next interval? wdyt?)